File: knowb/src/users/user_settings.py
from dataclasses import dataclass
from src.services import supabase


@dataclass
class UserPrompt:
    id: str
    prompt_texts: dict


async def get_user_prompt(user_id: str | None) -> UserPrompt:
    """Get the user's current prompt (or default), returning both ID and texts."""

    # Get prompt_id from user settings if it exists
    if user_id:
        settings = (
            supabase.from_("user_settings")
            .select("current_prompt_id")
            .eq("user_id", user_id)
            .single()
            .execute()
        )
        prompt_id = settings.data.get("current_prompt_id") if settings.data else None
    else:
        prompt_id = None

    # Get the prompt data (either user-selected or default)
    prompt_query = supabase.from_("prompts").select("id, prompt_texts")

    if prompt_id:
        prompt_query = prompt_query.eq("id", prompt_id)
    else:
        prompt_query = prompt_query.is_("user_id", None)

    result = prompt_query.single().execute()

    if not result.data:
        raise ValueError("No prompt found (neither user-selected nor default)")

    return UserPrompt(id=result.data["id"], prompt_texts=result.data["prompt_texts"])

File: knowf/src/app/settings/userSettings.ts
import { supabase } from "@/lib/supabase";

type PromptType = "pair";

export interface Prompt {
  id: string;
  user_id: string | null;
  name: string;
  prompt_type: PromptType;
  prompt_texts: {
    brainstorm_prompt: string;
    final_prompt: string;
  };
  is_active: boolean;
}

export interface UserSettings {
  user_id: string;
  current_prompt_id: string | null;
}

export async function ensureUserSettings(userId: string) {
  const { data } = await supabase
    .from("user_settings")
    .select()
    .eq("user_id", userId)
    .single();

  if (!data) {
    await supabase.from("user_settings").insert({ user_id: userId });
  }
}

export async function getCurrentPromptId(
  userId: string | null
): Promise<string> {
  if (userId) {
    const { data: settings } = await supabase
      .from("user_settings")
      .select("current_prompt_id")
      .eq("user_id", userId)
      .single();

    if (settings?.current_prompt_id) {
      return settings.current_prompt_id;
    }
  }

  // Fall back to default prompt
  const { data: defaultPrompt } = await supabase
    .from("prompts")
    .select("id")
    .is("user_id", null)
    .single();

  if (!defaultPrompt) {
    throw new Error("No default prompt found");
  }

  return defaultPrompt.id;
}

export async function getPromptTexts(promptId: string) {
  const { data: prompt, error } = await supabase
    .from("prompts")
    .select("prompt_texts")
    .eq("id", promptId)
    .single();

  if (error) throw error;
  if (!prompt) throw new Error("Prompt not found");

  return prompt.prompt_texts;
}

export async function getCurrentPrompt(userId: string) {
  const promptId = await getCurrentPromptId(userId);
  return getPromptTexts(promptId);
}
